[C10D] Add duration_ms to flight recorder

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114817

Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

Note: computing duration can lead to a hang due to calling cudaEventDuration when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang. Hence, we only query durations from the watchdog thread,
and it's possible during dump() call, some of the most recent
collectives durations won't have been computed yet at time of dump.  We
make this tradeoff to ensure that dump() itself will never hang.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225

wconstab committed:
WIP add duration to flight recorder

will be easier to finish this PR after
https://github.com/pytorch/pytorch/pull/114804 lands.

[ghstack-poisoned]

wconstab committed:
Update on "WIP add duration to flight recorder"


will be easier to finish this PR after
https://github.com/pytorch/pytorch/pull/114804 lands.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "WIP add duration to flight recorder"


will be easier to finish this PR after
https://github.com/pytorch/pytorch/pull/114804 lands.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "WIP add duration to flight recorder"


will be easier to finish this PR after
https://github.com/pytorch/pytorch/pull/114804 lands.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -439,6 +456,7 @@ struct NCCLTraceBuffer {
       if (completed) {
         r.state_ = "completed";
       }
+      r.duration_ = getDurationFromFirstEvent(*r.start_, *r.end_);
```
oops, i should probably put the getDuration up above in the if(completed) clause.

actually, does anyone know if its valid to measure the duration of a collective that crashed or was aborted?  cc @kwen2501

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1971,18 +1971,12 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
-      "getDuration can only be called after work is succeeded.")
-  return (*ncclStartEvents_)[0].elapsed_time((*ncclEndEvents_)[0]);
+  TORCH_CHECK(timingEnabled_, "getDuration only works if timing was enabled");
+  TORCH_CHECK(ncclStartEvents_, "getDuration only works if timing was enabled");
+  TORCH_CHECK(ncclEndEvents_, "getDuration only works if timing was enabled");
```
nit: do you want to update the error msg here?

wconstab replied:
i was intentionally lazy here. i think timing enabled --> start/end events should exist.  but i could make it more literally correct

fduwjj replied:
Well, I guess folks might want to change it later as a BE work anyway, lol.

fduwjj reviewed (approved):
LGTM

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -438,6 +455,7 @@ struct NCCLTraceBuffer {
       }
       if (completed) {
         r.state_ = "completed";
+        r.duration_ = getDurationFromFirstEvent(*r.start_, *r.end_);
```
If timing is disabled, then `r.start_` can be nullptr here and it will crash. I think we might want to turn on timing and start events if the flight recorder is enabled. That way we don't have all these different pathways where some information is not gathered.

wconstab replied:
yea, if possible lets enable timing by default. but i should make this safe regardless.  (IIRC there are checks _inside_ the getDuration fn but i probably forgot about the * or thought that it was always safe to deref the first layer.

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 kiukchung lucasllc

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 kiukchung lucasllc

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 kiukchung lucasllc

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 kiukchung lucasllc

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -475,16 +492,61 @@ struct NCCLTraceBuffer {
+      }
+    }
+
+    mutex_.unlock();
```
I think this needs to be a std::unique_lock with a call to `unlock` rather than a lock_guard. Otherwise a exception from getDurationFromFirstEvent will lead to a double-unlock of mutex_

zdevito replied:
`auto & entry` becomes invalid at this point. The vector can be resized.

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -475,16 +492,61 @@ struct NCCLTraceBuffer {
+    // Compute duration without without holding the lock, because
+    // cudaEventDuration() can hang, and we need to acquire the lock before we
+    // can dump(), which we never want to block.
+    if (can_compute_duration) {
```
Can't we put the if-statement around the locking logic?

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -475,16 +492,61 @@ struct NCCLTraceBuffer {
       entry.start_ = entry.end_ = nullptr;
+    } else {
+      LOG(INFO)
+          << "Failed retire_id " << *id
```
I wouldn't use the word Failed here. If we hit this, it just means that it was configured to no longer be interested in this event.

wconstab replied:
point taken. user doesn't care about this event anymore. (but its kind of unexpected / buggy (?) if we take so long to reacquire the lock that our whole buffer gets overwritten before we can compute durations and mark finished) -- worst case this could show up as durations are never working or are spotty in the part of the flight recorder you do dump

zdevito reviewed (commented):
Some minor comments, but a few are potential ub

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225 kiukchung lucasllc

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

Note: computing duration can lead to a hang due to calling cudaEventDuration when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang. Hence, we only query durations from the watchdog thread,
and it's possible during dump() call, some of the most recent
collectives durations won't have been computed yet at time of dump.  We
make this tradeoff to ensure that dump() itself will never hang.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -f "broken mergability check? (how do i know)"

wconstab committed:
Update on "[C10D] Add duration_ms to flight recorder"


Measures the duration of a collective operation using nccl start/end
events and includes this duration (in ms) in the flight recorder data.

duration_ms will be an optional field, since it only works when
timing is enabled.  Currently timing is enabled when flight recorder
is enabled, but this is not a strict requirement.  Duration is also
not available for collectives not in a completed state.

Note: computing duration can lead to a hang due to calling cudaEventDuration when
the cuda driver queue is full.

We don't ever want dump() api to hang, since we might want dump to help
debug a hang. Hence, we only query durations from the watchdog thread,
and it's possible during dump() call, some of the most recent
collectives durations won't have been computed yet at time of dump.  We
make this tradeoff to ensure that dump() itself will never hang.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
weird- this looks like it actually did merge. I see the PR on trunk.  closing manually.  cc @huydhn

huydhn commented:
Yeah, it was merged.  Trunk workflow was broken for few hours last Friday due to a permission issue (we revoked some AWS credentials), so this was landed during that windows.
